src/QsCompiler/Transformations/CallGraph/ConcreteCallGraphWalker.cs

```diff
@@ -0,0 +1,275 @@
+            {
+                var walker = new BuildGraph(graph);
+                var entryPointNodes = compilation.EntryPoints.Select(name => new ConcreteCallGraphNode(name, QsSpecializationKind.QsBody, TypeParameterResolutions.Empty));
+                foreach (var entryPoint in entryPointNodes)
+                {
+                    // Make sure all the entry points are added to the graph
+                    walker.SharedState.Graph.AddNode(entryPoint);
+                    walker.SharedState.RequestStack.Push(entryPoint);
+                }
+
+                var globals = compilation.Namespaces.GlobalCallableResolutions();
+                while (walker.SharedState.RequestStack.TryPop(out var currentRequest))
+                {
+                    // If there is a call to an unknown callable, throw exception
+                    if (!globals.TryGetValue(currentRequest.CallableName, out QsCallable currentCallable))
+                    {
+                        throw new ArgumentException($"Couldn't find definition for callable: {currentRequest.CallableName}");
+                    }
+
+                    var spec = GetSpecializationFromRequest(currentRequest, currentCallable);
+
+                    // The current request must be added before it is processed to prevent
+                    // self-references from duplicating on the stack.
+                    walker.SharedState.ResolvedNodeSet.Add(currentRequest);
+                    walker.SharedState.CurrentNode = currentRequest;
+                    walker.Namespaces.OnSpecializationImplementation(spec.Implementation);
+                }
+            }
+
+            /// <summary>
+            /// Returns the specialization that the request is referring to.
+            /// </summary>
+            private static QsSpecialization GetSpecializationFromRequest(ConcreteCallGraphNode request, QsCallable callable)
+            {
+                var relevantSpecs = callable.Specializations.Where(s => s.Kind == request.Kind);
+                var typeArgs = QsNullable<ImmutableArray<ResolvedType>>.Null;
+                if (callable.Signature.TypeParameters.Any())
+                {
+                    // Convert ParamResolutions from a dictionary to a type argument list
+                    typeArgs = QsNullable<ImmutableArray<ResolvedType>>.NewValue(callable.Signature.TypeParameters.Select(typeParam =>
+                    {
+                        if (typeParam is QsLocalSymbol.ValidName name)
+                        {
+                            if (request.ParamResolutions.TryGetValue(Tuple.Create(callable.FullName, name.Item), out var res))
+                            {
+                                return res;
+                            }
+                            else
+                            {
+                                throw new ArgumentException($"Couldn't resolve all type parameters for {callable.FullName}");
```

Comment:
I was able to get this exception to throw on a Q# program that compiles without errors:

```qsharp
function Foo<'a>(x : 'a) : Unit { }

function Bar<'a>(x : 'a) : 'a {
    return x;
}

@EntryPoint()
function Main() : Unit {
    Foo(Bar);
}
```

> System.ArgumentException: Couldn't resolve all type parameters for Graph.Bar

But I feel like this program shouldn't compile, actually...

Reply:
Simpler example:

```qsharp
function Foo<'a>() : Unit { }

@EntryPoint()
function Main() : Unit {
    let x = Foo();
}
```

> System.ArgumentException: Couldn't resolve all type parameters for Graph.Foo

Reply:
Here's an example where the type parameter actually does resolve to something concrete (`Int` and `Int -> Int`), but the concrete call graph still throws "Couldn't resolve all type parameters":

```qsharp
function Foo<'a>(x : 'a) : 'a {
    return x;
}

@EntryPoint()
function Main() : Unit {
    let x = (Foo(Foo))(0);
}
```

Reply:
My understanding is that none of these scenarios, even the last one, should be supported. It is odd to me that the call graph error is surfacing though, and maybe an issue should be filed to better protected against these cases. Either way that would be outside the scope of this PR.

Reply:
The first two are reasonable to forbid (they are both the same issue as you mention, namely that the return type of the callable is fully determined even when the callable is not concrete) - @samarsha Do you mind filing a bug? 
The last one however is perfectly possible to infer what the concrete callable(s) is/are. The third case I believe is fully supported throughout both the compiler and runtime, and we should keep it that way. However, the changes we are going to do to support [this task](https://github.com/microsoft/qsharp-compiler/issues/573) should actually make it easier to support this in the call graph. I am hence ok with adding a test for the third case that is currently skipped.

Reply:
> they are both the same issue as you mention, namely that the return type of the callable is fully determined even when the callable is not concrete

Is this the problem? There are useful functions with this property, for example `Length (a : 'T[]) : Int`.

Reply:
Ah right the last scenario is something we are looking into. I have added a unit test for that scenario that is, for now, skipped.

Reply:
It could be very useful to have a centralized place, like a design doc somewhere, to point people toward for information about generics in Q#, having information about what is currently supported, what we plan to support, what we have decided not to support, as well as having information about what scenarios are covered with type argument inference in Q#. @bettinaheim do we currently have something like this?

Reply:
The first one is probably OK to forbid as having an ambiguous type variable, since `Bar` is assigned to a parameter without its type being determined. I don't think we should forbid the second one, though, but maybe unused type variables should be removed from the AST before monomorphization.

Reply:
Removing unused type parameters before monomorphization could be a solution. However, let's take this discussion elsewhere as the details of this aren't in scope for this PR. @samarsha please file a github issue for this generics behavior you'd like to address, and we can continue this discussion there.